cmd/cstor-pool-mgmt/controller/pool-controller/handler.go

```diff
@@ -577,40 +578,39 @@ func (c *CStorPoolController) getDeviceIDs(csp *apis.CStorPool) ([]string, error
 
 func (c *CStorPoolController) reconcileVersion(csp *apis.CStorPool) (*apis.CStorPool, error) {
 	var err error
+	// the below code uses deep copy to have the state of object just before
+	// any update call is done so that on failure the last state object can be returned
 	if csp.VersionDetails.Status.Current != csp.VersionDetails.Desired {
+		cspObj := csp.DeepCopy()
 		if csp.VersionDetails.Status.State != apis.ReconcileInProgress {
-			csp.VersionDetails.Status.State = apis.ReconcileInProgress
-			csp.VersionDetails.Status.LastUpdateTime = metav1.Now()
-			csp, err = c.clientset.OpenebsV1alpha1().CStorPools().Update(csp)
+			cspObj.VersionDetails.Status = upgrade.SetPendingStatus(cspObj.VersionDetails.Status)
+			cspObj, err = c.clientset.OpenebsV1alpha1().CStorPools().Update(cspObj)
 			if err != nil {
-				return nil, err
+				return csp, err
 			}
 		}
-		if !isCurrentVersionValid(csp) {
-			return nil, errors.Errorf("invalid current version %s", csp.VersionDetails.Status.Current)
+		if !upgrade.IsCurrentVersionValid(cspObj.VersionDetails) {
+			return cspObj, errors.Errorf("invalid current version %s", cspObj.VersionDetails.Status.Current)
 		}
-		if !isDesiredVersionValid(csp) {
-			return nil, errors.Errorf("invalid desired version %s", csp.VersionDetails.Desired)
+		if !upgrade.IsDesiredVersionValid(cspObj.VersionDetails) {
+			return cspObj, errors.Errorf("invalid desired version %s", cspObj.VersionDetails.Desired)
 		}
-		path := upgradePath(csp)
+		path := upgrade.Path(csp.VersionDetails)
```

Comment:
even though csp.VersionDetails seems to be fine, should we consider passing `cspObj.VersionDetails`?

Reply:
Done